File: 00-Foundations/04-Design-Patterns/Behavioral/Strategy/strategy.ts
interface FeeStrategy {
  calculate(hours: number): number;
}

class HourlyFeeStrategy implements FeeStrategy {
  calculate(hours: number): number {
    return 10 * hours;
  }
}

class FlatDayRateStrategy implements FeeStrategy {
  calculate(_hours: number): number {
    return 50;
  }
}

class FreeFirstHourStrategy implements FeeStrategy {
  calculate(hours: number): number {
    return hours <= 1 ? 0 : 10 * (hours - 1);
  }
}

class ParkingTicket {
  // TS idiom: strategies are often passed as plain functions instead of
  // classes implementing an interface, when there's no extra state to
  // carry. Both are shown here — pick whichever fits the case study.
  constructor(private readonly strategy: FeeStrategy) {}

  calculateFee(hours: number): number {
    return this.strategy.calculate(hours);
  }
}

// Function-based Strategy — equally valid in TS, less ceremony when the
// "strategy" has no internal state of its own.
type FeeStrategyFn = (hours: number) => number;

function calculateWithStrategy(hours: number, strategy: FeeStrategyFn): number {
  return strategy(hours);
}

function demo(): void {
  const promoTicket = new ParkingTicket(new FreeFirstHourStrategy());
  console.log(promoTicket.calculateFee(3)); // 20

  console.log(calculateWithStrategy(3, (h) => 10 * h)); // 30, function-based strategy
}

demo();

File: 00-Foundations/04-Design-Patterns/Behavioral/Strategy/Strategy.cs
// Illustrates notes.md in this folder — the Strategy pattern.
//   dotnet run --project Runner strategy

namespace Foundations.Patterns.Behavioral.Strategy;

public interface IFeeStrategy
{
    decimal Calculate(int hours);
}

public class HourlyFeeStrategy : IFeeStrategy
{
    public decimal Calculate(int hours) => 10m * hours;
}

public class FlatDayRateStrategy : IFeeStrategy
{
    public decimal Calculate(int hours) => 50m; // capped, regardless of hours
}

public class FreeFirstHourStrategy : IFeeStrategy
{
    public decimal Calculate(int hours) => hours <= 1 ? 0m : 10m * (hours - 1);
}

// Depends only on the interface. Swapping pricing schemes never requires
// editing this class — new strategies are added, not this class modified.
public class ParkingTicket
{
    private readonly IFeeStrategy _strategy;

    public ParkingTicket(IFeeStrategy strategy)
    {
        _strategy = strategy;
    }

    public decimal CalculateFee(int hours) => _strategy.Calculate(hours);
}

public static class StrategyDemo
{
    public static void Run()
    {
        var promoTicket = new ParkingTicket(new FreeFirstHourStrategy());
        Console.WriteLine(promoTicket.CalculateFee(3)); // 20

        var regularTicket = new ParkingTicket(new HourlyFeeStrategy());
        Console.WriteLine(regularTicket.CalculateFee(3)); // 30
    }
}
